ipa-core/src/protocol/context/dzkp_validator.rs

```diff
@@ -0,0 +1,765 @@
+    fn convert(
+        x_left: &BitArray<[u8; 32], Lsb0>,
+        x_right: &BitArray<[u8; 32], Lsb0>,
+        y_left: &BitArray<[u8; 32], Lsb0>,
+        y_right: &BitArray<[u8; 32], Lsb0>,
+        prss_left: &BitArray<[u8; 32], Lsb0>,
+        prss_right: &BitArray<[u8; 32], Lsb0>,
+        z_right: &BitArray<[u8; 32], Lsb0>,
+    ) -> Self::UnverifiedFieldValues;
+}
+
+/// Validator Trait for DZKPs
+/// It is different from the validator trait since context outputs a `DZKPUpgradedContext`
+/// that is tied to a `DZKPBatch` rather than an `accumulator`.
+/// Function signature of `validate` is also different, it does not downgrade shares anymore
+#[async_trait]
+pub trait DZKPValidator<B: UpgradableContext> {
+    fn context(&self) -> B::DZKPUpgradedContext;
+
+    /// Allows to validate the current `DZKPBatch` and empties it. The associated context is then
+    /// considered safe until another multiplication is performed and thus new values are added
+    /// to `DZKPBatch`.
+    /// Is generic over `DZKPBaseFields`. Please specify a sufficiently large field for the current `DZKPBatch`.
+    async fn validate<DF: DZKPBaseField>(&self) -> Result<(), Error>;
+
+    /// `is_safe` checks that there are no remaining `UnverifiedValues` within the associated `DZKPBatch`
+    ///
+    /// ## Errors
+    /// Errors when there are `UnverifiedValues` left.
+    fn is_safe(&self) -> Result<(), Error>;
+
+    /// `get_chunk_size` returns the chunk size of the validator
+    fn get_chunk_size(&self) -> Option<usize>;
+
+    /// `seq_join` in this trait is a validated version of `seq_join`. It splits the input stream into `chunks` where
+    /// the `chunk_size` is defined within `validator`. Each `chunk` is a vector that is independently
+    /// verified using `validator.validate()`, which uses DZKPs. Once the validation fails,
+    /// the output stream will return that the stream is done.
+    ///
+    fn seq_join<'st, S, F, O, DF>(
+        &'st self,
+        source: S,
+    ) -> impl Stream<Item = Result<O, Error>> + 'st
+    where
+        S: Stream<Item = F> + Send + 'st,
+        F: Future<Output = O> + Send + 'st,
+        O: Send + Sync + Clone + 'static,
+        DF: DZKPBaseField,
+    {
+        // chunk_size is undefined in the semi-honest setting, set it to 10, ideally it would be 1
+        // but there is some overhead
+        let chunk_size = self.get_chunk_size().unwrap_or(10usize);
+        seq_join::<'st, S, F, O>(self.context().active_work(), source)
+            .chunks(chunk_size)
+            .then(move |chunk| self.validate::<DF>().map_ok(|()| chunk))
+            .try_flatten_iters()
+    }
+}
+
+pub struct SemiHonestDZKPValidator<'a, B: ShardBinding> {
+    context: SemiHonestDZKPUpgraded<'a, B>,
+}
+
+impl<'a, B: ShardBinding> SemiHonestDZKPValidator<'a, B> {
+    pub(super) fn new(inner: Base<'a, B>) -> Self {
+        // This is inconsistent with the malicious new() which calls dzkp_upgrade
```

Comment:
It feels like you want to do something about it but I am not sure what exactly

Reply:
I remove the comment and just keep it consistent with the previous framework